src/fasta.rs

```diff
@@ -1,209 +1,932 @@
+    /// and all values for other settings
+    fn assert_parse_with_allow_preceding_comment(
+        s: &str,
+        allow_preceding_comment: bool,
+        expected: Vec<FastaRecord<String>>,
+    ) {
+        assert_parse!(
+            s,
+            FastaParser::<String>::new(FastaParseSettings {
+                concatenate_headers: true,
+                allow_preceding_comment
+            }),
+            expected
+        );
+        assert_parse!(
+            s,
+            FastaParser::<String>::new(FastaParseSettings {
+                concatenate_headers: false,
+                allow_preceding_comment
+            }),
+            expected
+        );
+    }
+
     #[test]
     fn test_empty_fasta() {
-        let r = SimpleFastaParser::parse("".as_bytes()).unwrap();
-        assert_eq!(r.len(), 0);
+        assert_parse_with_all_settings("", vec![]);
     }
 
     #[test]
     fn test_empty_fasta_with_newlines() {
-        let r = SimpleFastaParser::parse("\n\n".as_bytes()).unwrap();
-        assert_eq!(r.len(), 0);
+        assert_parse_with_all_settings("\n\n", vec![]);
     }
 
     #[test]
     fn test_empty_fasta_with_many_newlines() {
-        let r = SimpleFastaParser::parse("  \n\n \n  \r\r \r\n  \r \n".as_bytes()).unwrap();
-        assert_eq!(r.len(), 0);
+        assert_parse_with_all_settings("  \n\n \n  \r\r \r\n  \r \n", vec![]);
     }
 
     #[test]
     fn test_fasta_with_no_content() {
-        let r = SimpleFastaParser::parse(">Virus\n".as_bytes()).unwrap();
-        assert_eq!(r.len(), 1);
-        assert_eq!(r[0], ("Virus".to_string(), "".to_string()));
+        assert_parse_with_all_settings(
+            ">Virus\n",
+            vec![FastaRecord {
+                header: "Virus".to_string(),
+                contents: "".to_string(),
+                line_range: (1, 2),
+            }],
+        );
     }
 
     #[test]
     fn test_fasta_with_empty_content() {
-        let r = SimpleFastaParser::parse(">Virus\n\n".as_bytes()).unwrap();
-        assert_eq!(r, vec![("Virus".to_string(), "".to_string())]);
+        assert_parse_with_all_settings(
+            ">Virus\n\n",
+            vec![FastaRecord {
+                header: "Virus".to_string(),
+                contents: "".to_string(),
+                line_range: (1, 3),
+            }],
+        );
+    }
+
+    #[test]
+    fn test_comment_fasta_with_allow_preceding_comment() {
+        assert_parse_with_allow_preceding_comment(
+            "this is a file comment\n@author is foo",
+            true,
+            vec![],
+        );
     }
 
     #[test]
-    fn test_fasta_with_stuff_no_header() {
-        let r =
-            SimpleFastaParser::parse("// this is a file comment\n@author is foo\n\n".as_bytes())
-                .unwrap();
-        assert_eq!(r.len(), 0);
+    fn test_comment_fasta_with_no_allow_preceding_comment() {
+        assert_parse_with_allow_preceding_comment(
```

Comment:
I don't think that this behavior is correct for the default case.

Reply:
Could you elaborate? Looking at it, the setting says "don't allow ignoring comments, parse it as a headerless record" and that's what the code seems to be doing, but I might be missing something (or the name might be misleading?)

Reply:
So, this test variant is testing a situation in which we do not allow preceding comments.

The existing code ignores anything that precedes the first header. In this case, everything.

I don't like the idea of assuming that anything without a header is some content.

If there is no `>`, it should not parse.

Reply:
Ohh, I see what you mean -- only allow a comment if there's at least one record. Hmm.

I guess my feeling is, we allow parsing an empty file, right? And comment with no records = an empty file? Like, if you had some system that spit out a comment like:

```
// sequenced-on: 2022-07-30
```

on every record, and some sequencing result came back empty, then you'd *want* to parse that as an empty file. And if you don't want comments, you can use `strict()` mode instead. But I see what you mean that this could bite you if you expected comments on non-empty files, but accidentally omit a header and then lose a record on something that shouldn't be an empty file. Argh. FASTA is a sucky format :-(

Reply:
(We could introduce an option like "comments must start with this prefix" to prevent this, but that's non-standard -- if hewing to the FASTA "standard" even makes sense! :-p )

Reply:
> So, this test variant is testing a situation in which we do not allow preceding comments.
> 
> The existing code ignores anything that precedes the first header. In this case, everything.
> 
> I don't like the idea of assuming that anything without a header is some content.
> 
> If there is no `>`, it should not parse.

Nope, disagree, for security reasons, which is why I specified this behavior in the API.

Here's the problem:  We can't be sure that some provider, or some benchtop, *won't* accept a sequence without a header.  And if anyone in the world does that, it provides the most trivial bypass possible for screening:  just never include a header in your order, and we won't screen it, but the provider will make it.  And if some provider insists on seeing *some* header but still considers stuff before the header as something to be manufactured, then a malicious customer submits "malicious stuff, header, benign stuff" and still escapes screening.

This way, we put the onus, both technical and legal, directly on the provider who is giving us the task of screening.  If they consider stuff before a header to be a comment, then they are required to remove it before passing it to us.  That guarantees that the behavior they think they want and the behavior they give to us will match.  They are free to either ignore it (and remove it before handing it to us) or complain to the customer and make *them* remove it, but anything they pass to us *will* be screened, and we aren't going to try to heuristicate either whether pre-any-header stuff is a comment or is valid content:  by treating it all as content, we force the provider or their UI to make that choice, which is a trivial amount of programming for them either way.

Reply:
Here is my immediate thought. The provider still has to somehow match the result of the screening to the order, right? For each sequence you should verify whether it is a hazard or not. Not having any reply for a sequence is clearly a problem that someone would notice, no?

Reply:
I also just noticed some of the upthread commentary on this re comments.

The deal is that anything which is not a sequence is a comment in FASTA.  That means any line which begins with > or ; (though ; is semi-obsolete).  Anything else is parse error.  By definition, at least as far as we're concerned, comments are *opaque blobs" with no semantic content whatsoever, because there is no standard about what goes in there.  Particular consumers of such files, such as NCBI, may have standards for what *they* accept (since in NCBI's case they're using them to create database keys for submissions), but in general there's just no way at all to predict what's going on and no enforcement even if there *was* a standard, which there isn't.  All we're doing with preserving these > or ; lines is making it slightly easier to make a UI that tells the customer where in the file we're objecting to something.  That's it.  So we really don't care about comments at all.  And the only fly in that ointment is the practice *in some FASTA files* of assuming that anything before the first > is also ignored.  We're enabling that with a switch, but we ourselves never intend to enable that for screening---you have something that could plausibly be a sequence before the first > and you don't want it synthesized or screened, then you put a ; or a > in front of each line of it.  Or if you're a provider and you want to allow that, *you* take it out before handing it to us.  Then it's clear who's responsible if you, as a provider, deleted that stuff before you gave it to us but then went ahead and made it anyway.

Reply:
(Also the way to deal with "you accidentally created a null record and didn't notice it" is not to overload the commenting/screening system into pointing out your error.  Either you as a customer need to run QA first on what you're ordering, or the provider may have its own rules/helpful hints and tells you, but that's not our job.  And it can't be, because valid FASTA files exist in the wild with multiple consecutive header lines and we can't tell that from an error.)

Reply:
> (We could introduce an option like "comments must start with this prefix" to prevent this, but that's non-standard -- if hewing to the FASTA "standard" even makes sense! :-p )

No, let's not; nobody will do this. :)  What they do instead is to just pack all kinds of trash into the > line, including random lab notes like "sequencing added ACCTTGGACA to the previous record" and all sorts of other random asides.  If they want something more sophisticated, they should be using tools that maintain their sequences in a DB (for example) and then spit out valid FASTA files with the extra info suppressed, especially if that info is just going into the maw of a synth, which is certainly not reading those comments anyway.  Trying to put all this metadata into the FASTA file itself is like (true story!) what I saw once where the exec dir  of a ~$1M/yr nonprofit thought the right way to issue PO's was to just create them directly as PDF's from a template and was amazed to hear that actual software existed in the world which tracked all that stuff in a DB and could then issue PO's & invoices & so forth downstream on demand.